pulsar-impl/src/main/java/com/datastax/oss/sink/pulsar/CassandraSinkTask.java

```diff
@@ -189,28 +199,22 @@ public void write(Record<T> record) throws Exception {
         log.info("raw {}", rawvalue);
       }
     }
-    int number;
-    synchronized (this) {
-      incomingList.add(record);
-      number = incomingList.size();
-    }
-    if (number == batchSize) {
+
+    incomingQueue.put(record); // blocking call to slow down the topic consumer
+    if (incomingQueue.size() >= batchSize) {
```

Comment:
`>=` instead of `=` to handle edge cases where the incomingQueueSize is already larger than the batch size when the code path is triggered - this could happen under high load

Reply:
this will submit multiple tasks to the executor. e.g. by the time .put() above blocks at 2*batchSize records you'll have batchSize of pending flush tasks submitted. `==` should be enough given that flush drains all the tasks / drops incomingQueue to zero size

Reply:
after line 204 reads exactly a batch size (the `==` case) and before the flush thread is scheduled and begins execution, nothing prevents the "put" call to accumulate extra records (in corner cases it could those extra records could be larger than a single batch). If that happens, the queue size will be past the batch size. It is correct that extra tasks will be submitted by the >= approach, but they will be ignored as only one execution is allowed.

another approach is to put the flush code inside a for loop (most of the time it will kick in once only), to drain all available batches and not wait for the next trigger if a second batch is already due. 

having said that, using == (even with the corner case I mentioned) is safe because of the scheduled flush task that will clean any leftovers (if any).